### PR TITLE
fix(router): use appRootComponentToken to get root route configs

### LINKED ISSUE #1947: Router error: Cannot instantiate Token(ComponentRef) synchronously.
I've built Angular2 and the router from src and I get an error `Cannot instantiate Token(ComponentRef) synchronously. It is provided as a promise! (RouterOutlet -> Router -> Token(ComponentRef))`

The code I use:

```typescript
/// <reference path="../typings/angular2/angular2.d.ts" />

import {Component, View, bootstrap} from 'angular2/angular2';
import {Router, RouterOutlet, RouteConfig, routerInjectables, RouterLink} from 'angular2/router';

@Component({
	selector: 'app'
})
@View({
	template: '<router-outlet></router-outlet>',	
	directives: [ RouterOutlet ]
})	
@RouteConfig([
	{ path: '/', component: Test, as: 'test' }
])
class App { }

@Component({
	selector: 'test'		
})	
@View({
	template: 'test'
})
class Test { }

bootstrap(App, [routerInjectables]);
```

### PR DESCRIPTION
**Pull request by btford**
Closes #1947 

/cc @rkirov @tbosch

### CODE REVIEW

File: modules/angular2/test/router/router_integration_spec.js [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+      testBindings = [
+        routerInjectables,
+        bind(Location).toClass(SpyLocation),
+        bind(DOCUMENT_TOKEN).toValue(fakeDoc)
+      ];
+    });
+
+    it('should support bootstrap a simple app', inject([AsyncTestCompleter], (async) => {
+      bootstrap(AppCmp, testBindings).then((refs) => {
+        expect(el).toHaveText('outer { matias }');
```
Comment (btford) [inline modules/angular2/test/router/router_integration_spec.js]: This assertion currently fails; I need to get ahold of the root router and subscribe to it to know when it finishes navigating. How can I do that with a bootstrapped app?

File: modules/angular2/src/core/application_tokens.js [hunk 2]
```diff
@@ -1,3 +1,4 @@
 import {OpaqueToken} from 'angular2/di';
 
 export var appComponentRefToken:OpaqueToken = new OpaqueToken('ComponentRef');
+export var appRootComponentToken:OpaqueToken = new OpaqueToken('RootComponent');
```
Comment (btford) [inline modules/angular2/src/core/application_tokens.js]: Is this a good name? @tbosch?

Comment (vicb) [inline modules/angular2/src/core/application_tokens.js]: `appComponentTypeToken ... ('AppComponentType')` ?

Comment (tbosch) [inline modules/angular2/src/core/application_tokens.js]: Yes, agree with @vicb: should contain `Type`

File: modules/angular2/test/router/router_integration_spec.js (hunk 1 above)
Comment (tbosch) [inline modules/angular2/test/router/router_integration_spec.js]: Couldn't you just inject the `RootRouter`? You probably need to to 
`bind(RootRouter).toClass(RootRouter), bind(Router).toAlias(RootRouter)` in `application.js`

File: modules/angular2/src/core/application_tokens.js (hunk 2 above)
Comment (btford) [inline modules/angular2/src/core/application_tokens.js]: :+1:

File: modules/angular2/src/router/router_outlet.js [hunk 3]
```diff
@@ -58,10 +60,11 @@ export class RouterOutlet {
   activate(instruction:Instruction): Promise {
     // if we're able to reuse the component, we just have to pass along the instruction to the component's router
     // so it can propagate changes to its children
-    if (instruction.reuse && isPresent(this._childRouter)) {
+    if ((instruction == this._currentInstruction) || instruction.reuse && isPresent(this._childRouter)) {
```
Comment (btford) [inline modules/angular2/src/router/router_outlet.js]: I need a unit test to cover this behavior as well; for some reason I'm just now encountering a case where the component gets activated twice.

But in the interest of not blocking the release can we merge this, and I'll promise to follow up later with a proper unit test.

Comment (rkirov) [inline modules/angular2/src/router/router_outlet.js]: add a todo

**Comment by btford** [comment]
Landed as 791caf003745e00460055c8a6a4c704a601931d8.

File: modules/angular2/src/router/router_outlet.js (hunk 3 above)
Comment (btford) [inline modules/angular2/src/router/router_outlet.js]: Ah, missed this. I added an issue to track it instead: https://github.com/angular/angular/issues/2023

**Pull request closed without merging**